Ported string.c to string.rs

waylon531 opened the pull request:
Ported string.c to rust. All of the string functions have been ported
over, they appear to work with usertest but they would be good
candidates for unit testing. The mem{cpy,cmp,mov} functions were not
ported over as they are already included with Rust's compiler-builtins
crate.

This fixes issue #32

squidboylan commented on src/string.rs:
```diff
@@ -0,0 +1,90 @@
+#[no_mangle]
+pub extern "C" fn strncmp(s: *const c_char, 
+                          t: *const c_char, 
+                          n: isize) 
```
In the C code `n` is a `uint` should this be `n: usize`? If it is then you dont have to check `n <= 0`

waylon531 replied:
Resolved in the latest commit. Thanks for the review!

waylon531 commented on src/string.rs:
```diff
@@ -0,0 +1,90 @@
+        //Return the difference with unmatching characters
+        if l != r { return *l as isize - *r as isize }
+        //And return 0 if we reached the end of the string
+        else if *l == 0 {return 0}
```
Welp, forgot to take that part out after I rewrote it to use `take_while`. This should be fixed in the most recent commit.

waylon531 commented:
Thanks for looking over my code! I ended up fixing the code up in the most recent commit, 91ef10f. @connorkuehl what's your preference on squashing these commits together?

squidboylan commented on src/string.rs:
```diff
@@ -0,0 +1,85 @@
+
+    //This will give weird results if you pass in a negative number
+    let s_slice = unsafe {slice::from_raw_parts(s,n as usize)};
+    let t_slice = unsafe {slice::from_raw_parts(t,n as usize)};
```
I'm still being a bit nit picky but you dont need `n as usize` here anymore since `n` is now a `usize`.

squidboylan commented:
IMO it makes git history easier to read and deal with (for rebasing or reverting or whatever).

waylon531 commented:
I fixed a bug in strcmp and usertests are passing now and I cleaned up the other functions as well to be closer to the original C code. I ran `rustfmt` on the code so it has now been formatted correctly, if the changes look good to you I'll squash all of them together so it's ready for merging

connorkuehl commented:
Thank you so much for the awesome work!
